pgmpy/inference/mplp.py

```diff
@@ -0,0 +1,502 @@
+
+        # Check if this is the best assignment till now
+        if self.best_int_objective < integer_value:
+            self.best_int_objective = integer_value
+            self.best_assignment = decoded_result_assignment
+
+    def _is_converged(self, dual_threshold=None, integrality_gap_threshold=None):
+        """
+        This method checks the integrality gap to ensure either:
+            * we have found a near to exact solution or
+            * stuck on a local minima.
+
+        Parameters
+        ---------
+        dual_threshold: double
+                        This sets the minimum width between the dual objective decrements. If the decrement is lesser
+                        than the threshold, then that means we have stuck on a local minima.
+
+        integrality_gap_threshold: double
+                                   This sets the threshold for the integrality gap below which we say that the solution
+                                   is satisfactory.
+
+        Reference:
+        code presented by Sontag in 2012 here: http://cs.nyu.edu/~dsontag/code/README_v2.html
+        """
+        # Find the new objective after the message updates
+        new_dual_lp = sum([max(self.objective[obj].values) for obj in self.objective])
+
+        # Update the dual_gap as the difference between the dual objective of the previous and the current iteration.
+        self.dual_gap = abs(self.dual_lp - new_dual_lp)
+
+        # Update the integrality_gap as the difference between our best result vs the dual objective of the lp.
+        self.integrality_gap = abs(self.dual_lp - self.best_int_objective)
+
+        # As the decrement of the dual_lp gets very low, we assume that we might have stuck in a local minima.
+        if dual_threshold and self.dual_gap < dual_threshold:
+            return True
+        # Check the threshold for the integrality gap
+        elif integrality_gap_threshold and self.integrality_gap < integrality_gap_threshold:
+            return True
+        else:
+            self.dual_lp = new_dual_lp
+            return False
+
+    def find_triangles(self):
+        """
+        Finds all the triangles present in the given model
+
+        Examples
+        --------
+        >>> from pgmpy.models import MarkovModel
+        >>> from pgmpy.factors import Factor
+        >>> from pgmpy.inference import Mplp
+        >>> mm = MarkovModel()
+        >>> mm.add_nodes_from(['x1', 'x2', 'x3', 'x4', 'x5', 'x6', 'x7'])
+        >>> mm.add_edges_from([('x1', 'x3'), ('x1', 'x4'), ('x2', 'x4'),
+        ...                    ('x2', 'x5'), ('x3', 'x6'), ('x4', 'x6'),
+        ...                    ('x4', 'x7'), ('x5', 'x7')])
+        >>> phi = [Factor(edge, [2, 2], np.random.rand(4)) for edge in mm.edges()]
+        >>> mm.add_factors(*phi)
+        >>> mplp = Mplp(mm)
+        >>> mplp.find_triangles()
+        """
+        return [clique for clique in nx.find_cliques(self.model) if len(clique) == 3]
```

Comment:
Use ```filter``` for this

```python
return list(filter(lambda x: len(x) == 3, nx.find_cliques(self.model)])
```

This would be much faster as compared to list comprehension with conditions.